cpp/src/arrow/python/python_to_arrow.cc

```diff
@@ -329,985 +302,602 @@ struct ValueConverter<DurationType> {
-      ARROW_ASSIGN_OR_RAISE(this->string_view_,
-                            ValueConverter<Type>::FromPython(obj, &is_utf8));
-      if (!is_utf8) {
-        ++binary_count_;
-      }
+      ARROW_ASSIGN_OR_RAISE(auto converted,
+                            PyValue::Convert(this->value_type_, this->options_, value));
+      return this->value_builder_->Append(converted);
     }
-    return this->AppendString(this->string_view_);
   }
+};
 
-  Status GetResult(std::shared_ptr<ChunkedArray>* out) override {
-    RETURN_NOT_OK(SeqConverter::GetResult(out));
-
-    // If we saw any non-unicode, cast results to BinaryArray
-    if (binary_count_) {
-      // We should have bailed out earlier
-      DCHECK(!Strict);
-      auto binary_type = TypeTraits<typename Type::PhysicalType>::type_singleton();
-      return (*out)->View(binary_type).Value(out);
+template <typename U>
+class PyDictionaryConverter<U, enable_if_has_string_view<U>>
+    : public DictionaryConverter<U, PyConverter> {
+ public:
+  Status Append(PyObject* value) override {
+    if (PyValue::IsNull(this->options_, value)) {
+      return this->value_builder_->AppendNull();
+    } else {
+      ARROW_ASSIGN_OR_RAISE(auto view,
+                            PyValue::Convert(this->value_type_, this->options_, value));
+      return this->value_builder_->Append(util::string_view(view.bytes, view.size));
     }
-    return Status::OK();
   }
-
- protected:
-  int64_t binary_count_;
 };
 
-// ----------------------------------------------------------------------
-// Convert lists (NumPy arrays containing lists or ndarrays as values)
-
 // If the value type does not match the expected NumPy dtype, then fall through
 // to a slower PySequence-based path
-#define LIST_FAST_CASE(TYPE, NUMPY_TYPE, ArrowType)            \
-  case Type::TYPE: {                                           \
-    if (PyArray_DESCR(arr)->type_num != NUMPY_TYPE) {          \
-      return value_converter_->Extend(obj, value_length);      \
-    }                                                          \
-    return AppendNdarrayTypedItem<NUMPY_TYPE, ArrowType>(arr); \
+#define LIST_FAST_CASE(TYPE_ID, TYPE, NUMPY_TYPE)               \
+  case Type::TYPE_ID: {                                         \
+    if (PyArray_DESCR(ndarray)->type_num != NUMPY_TYPE) {       \
+      return Extend(this->value_converter_.get(), value, size); \
+    }                                                           \
+    return AppendNdarrayTyped<TYPE, NUMPY_TYPE>(ndarray);       \
   }
 
 // Use internal::VisitSequence, fast for NPY_OBJECT but slower otherwise
-#define LIST_SLOW_CASE(TYPE)                            \
-  case Type::TYPE: {                                    \
-    return value_converter_->Extend(obj, value_length); \
+#define LIST_SLOW_CASE(TYPE_ID)                               \
+  case Type::TYPE_ID: {                                       \
+    return Extend(this->value_converter_.get(), value, size); \
   }
```

Comment:
Since these macros are only used in PyListConverter::AppendNdarray, please #define and #undef them inside that scope